domain/miningmanager/mempool/mempool.go

```diff
@@ -1,36 +1,915 @@
+// rules, orphan transaction handling, and insertion into the memory pool.
+//
+// It returns a slice of transactions added to the mempool. When the
+// error is nil, the list will include the passed transaction itself along
+// with any additional orphan transaactions that were added as a result of
+// the passed one being accepted.
+//
+// This function is safe for concurrent access.
+func (mp *mempool) ValidateAndInsertTransaction(tx *consensusexternalapi.DomainTransaction, allowOrphan bool) error {
+	log.Tracef("Processing transaction %s", hashserialization.TransactionID(tx))
+
+	// Protect concurrent access.
+	mp.mtx.Lock()
+	defer mp.mtx.Unlock()
+
+	// Potentially accept the transaction to the memory pool.
+	missingParents, txD, err := mp.maybeAcceptTransaction(tx, true)
+	if err != nil {
+		return err
+	}
+
+	if len(missingParents) == 0 {
+		// Accept any orphan transactions that depend on this
+		// transaction (they may no longer be orphans if all inputs
+		// are now available) and repeat for those accepted
+		// transactions until there are no more.
+		newTxs := mp.processOrphans(tx)
+		acceptedTxs := make([]*txDesc, len(newTxs)+1)
+
+		// Add the parent transaction first so remote nodes
+		// do not add orphans.
+		acceptedTxs[0] = txD
+		copy(acceptedTxs[1:], newTxs)
+
+		return nil
+	}
+
+	// The transaction is an orphan (has inputs missing). Reject
+	// it if the flag to allow orphans is not set.
+	if !allowOrphan {
+		// Only use the first missing parent transaction in
+		// the error message.
+		//
+		// NOTE: RejectDuplicate is really not an accurate
+		// reject code here, but it matches the reference
+		// implementation and there isn't a better choice due
+		// to the limited number of reject codes. Missing
+		// inputs is assumed to mean they are already spent
+		// which is not really always the case.
+		str := fmt.Sprintf("orphan transaction %s references "+
+			"outputs of unknown or fully-spent "+
+			"transaction %s", hashserialization.TransactionID(tx), missingParents[0])
+		return txRuleError(RejectDuplicate, str)
+	}
+
+	// Potentially add the orphan transaction to the orphan pool.
+	return mp.maybeAddOrphan(tx)
+}
+
+// Count returns the number of transactions in the main pool. It does not
+// include the orphan pool.
+//
+// This function is safe for concurrent access.
+func (mp *mempool) Count() int {
+	mp.mtx.RLock()
+	defer mp.mtx.RUnlock()
+	count := len(mp.pool)
+
+	return count
+}
+
+// ChainedCount returns the number of chained transactions in the mempool. It does not
+// include the orphan pool.
+//
+// This function is safe for concurrent access.
+func (mp *mempool) ChainedCount() int {
+	mp.mtx.RLock()
+	defer mp.mtx.RUnlock()
+	return len(mp.chainedTransactions)
+}
+
+// MiningDescs returns a slice of mining descriptors for all the transactions
+// in the pool.
+//
+// This is part of the mining.TxSource interface implementation and is safe for
```

Comment:
It seems like mining.TxSource is no longer a thing, if so - remove this comment (except the part about concurrent use)